sktime/forecasting/trend/_function_forecaster.py

```diff
@@ -0,0 +1,133 @@
+# !/usr/bin/env python3 -u
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+"""Implements FunctionForecaster."""
+
+__author__ = ["benheid"]
+__all__ = ["FunctionForecaster"]
+
+import numpy as np
+import pandas as pd
+from scipy.optimize import curve_fit
+
+from sktime.forecasting.base import BaseForecaster, ForecastingHorizon
+
+
+class FunctionForecaster(BaseForecaster):
+    """The FunctionForecaster uses a function that is fitted using scipy curve_fit.
+
+    Parameters
+    ----------
+    function: Callable
+        The function that should be fitted and used to make forecasts.
+    initial_params : list
+        The initial parameters of the functions that should be used for fitting
+
+
+    Examples
+    --------
+    >>> from sktime.forecasting.trend import FunctionForecaster
+    >>> from sktime.datasets import load_airline
+    >>> y = load_airline()
+    >>> def linear_function(x, a, b):
+    ...     return a * x + b
+    >>> forecaster = FunctionForecaster(function=linear_function,
+    ...                                 initial_params=[1, 1])
+    >>> forecaster.fit(y)
+    FunctionForecaster(...)
+    >>> y_pred = forecaster.predict(fh=[1, 2, 3])
+    """
+
+    _tags = {
+        "scitype:y": "univariate",
+        "y_inner_mtype": "pd.Series",
+        "X_inner_mtype": "pd.DataFrame",
+        "ignores-exogeneous-X": True,
+        "requires-fh-in-fit": False,
+    }
+
+    def __init__(self, function, initial_params):
+        self.function = function
+        self.initial_params = initial_params
+        super().__init__()
+
+    def _fit(self, y, X=None, fh=None):
+        """Fit forecaster to training data.
+
+        Writes to self:
+            Sets fitted model attributes ending in "_".
+
+        Parameters
+        ----------
+        y : pd.DataFrame
+            time series to which to fit the forecaster.
+        fh : guaranteed to be ForecastingHorizon or None, optional (default=None)
+            The forecasting horizon with the steps ahead to predict.
+            Ignored in fit
+        X : pd.DataFrame, optional (default=None)
+            Ignored in fit
+
+        Returns
+        -------
+        self : reference to self
+        """
+        t = ForecastingHorizon(y.index, is_relative=False).to_relative(self.cutoff)
+        self.params_ = curve_fit(
+            self.function, np.array(t), y.values, self.initial_params
+        )
```

Comment:
I wanted your opinion on whether you'd think it would be better if we could add some external validations that raise errors which provide more clarity in the context of forecasting. I reckon that `curve_fit` must be running its own validations but I'm not sure if the errors they provide would be useful for the users of `FunctionForecaster`.

Reply:
According to the scipy docs, three errors can be raised:

* ValueError, if either ydata or xdata contain NaNs, or if incompatible options are used.
* RuntimeError, if the least-squares minimization fails.
* OptimizeWarning, if covariance of the parameters can not be estimated.

I am unsure how the forecasting context can provide more useful information here..